File: AlgoExpert/Stacks/Medium/Balanced Brackets.java
import java.util.*;

class Program {
  public static boolean balancedBrackets(String str) {
    // Write your code here.
    String openingBrackets = "({[";
    String closingBrackets = ")}]";
    Stack<Character> stack = new Stack<>();

    HashMap<Character, Character> matchingBrackets = new HashMap<>();
    matchingBrackets.put(')', '(');
    matchingBrackets.put('}', '{');
    matchingBrackets.put(']', '[');
    
    for (int i = 0; i < str.length(); i++) {
      char current = str.charAt(i);
      boolean isOpeningBracket = openingBrackets.indexOf(current) != -1;
      boolean isClosingBracket = closingBrackets.indexOf(current) != -1;
      
      if (isOpeningBracket) {
        stack.push(current);
      } else if (isClosingBracket) {
        if (stack.isEmpty() || stack.peek() != matchingBrackets.get(current)) return false;
        else stack.pop();
      }
    }
      
    return stack.isEmpty();
  }
}

File: Python/AlgoExpert/Stacks/Medium/balancedBrackets.py
def balancedBrackets(string):
    # Write your code here.
    bracket_map = {'(': ')', '{': '}', '[': ']'}
    opening_brackets = '({['
    closing_brackets = ')}]'
    
    curr_open = []
    for i in range(len(string)):
        if string[i] in opening_brackets:
            curr_open.append(string[i])
        elif string[i] in closing_brackets:
            if len(curr_open) == 0:
                return False
            else:
                last_open = curr_open.pop()
                if bracket_map[last_open] != string[i]:
                    return False
        
    return len(curr_open) == 0
    
                
